### PR TITLE
Chaining implementation with prototype

### PR DESCRIPTION
**Pull request by tshemsedinov**
Alternative implementation with prototypes for performance.
@nechaido we need to compare with promise and functor implementation.

### CODE REVIEW

File: lib/chain.js [hunk 1]
```diff
@@ -2,80 +2,90 @@
...
-  class ArrayChain {
-
-    constructor(array) {
-      this._promise = Promise.resolve(array);
-    }
-
-    then(fn) {
-      return this._promise.then(fn);
+  const async = op => {
+    switch (op) {
```
Comment (nechaido) [inline lib/chain.js]: Wouldn't it be better to use hashmap or array here?

File: lib/chain.js [hunk 2]
```diff
@@ -2,80 +2,90 @@
...
-
-    series(fn) {
-      this._update((array, resolve, reject) => {
-        api.metasync.series(array, fn, resolver(resolve, reject));
-      });
-      return this._promise;
-    }
+  function ArrayChain(array) {
+    this.array = array;
+    this.chain = [];
```
Comment (nechaido) [inline lib/chain.js]: Maybe linked list would be better here(needs testing)

File: lib/chain.js (hunk 1 above)
Comment (tshemsedinov) [inline lib/chain.js]: `switch/case` gives more performance

File: lib/chain.js (hunk 2 above)
Comment (tshemsedinov) [inline lib/chain.js]: Even theoretically array gives more performance, but in V8 arrays are highly optimized, so better you will test it and I expect arrays to be best.

**Comment by nechaido** [comment]
@tshemsedinov should I resolve conflicts and merge it?

**Comment by tshemsedinov** [comment]
@nechaido yes:

1. Use fetch instead of then/catch, see [#issuecomment-293739342](https://github.com/metarhia/MetaSync/issues/72#issuecomment-293739342) in 
2. Resolve conflicts
3. Squash b4567fdf40362f57eb152d5a42e3fa218088a182 and 3d3262f77ed59d61d159622670ead3fa2dc72bbb
4. Land in master

**Comment by nechaido** [comment]
@tshemsedinov I`ll wait for review on fetch implementation and squash and merge after if everything is ok

**Comment by tshemsedinov** [comment]
@nechaido I am working on this branch right now, including #80 and #81

File: lib/chain.js [hunk 3]
```diff
@@ -22,34 +22,28 @@ module.exports = (api) => {
...
 
     if (!item.op) {
-      if (err) {
-        throw err;
-      } else {
-        return;
-      }
+      if (err) throw err;
+      else return;
     }
```
Comment (tshemsedinov) [inline lib/chain.js]: Just shorter syntax

File: lib/chain.js [hunk 4]
```diff
@@ -22,34 +22,28 @@ module.exports = (api) => {
...
 
     if (item.op === 'fetch') {
       item.fn(err, this.array);
-      return this.execute();
```
Comment (tshemsedinov) [inline lib/chain.js]: Before #80 chaining after fetch have no sense, but we will implement it later

File: lib/chain.js [hunk 5]
```diff
@@ -22,34 +22,28 @@ module.exports = (api) => {
...
+      return;
     }
 
     const op = async(item.op);
-    if (!op) return this.execute();
```
Comment (tshemsedinov) [inline lib/chain.js]: This case will newer happen

File: lib/chain.js [hunk 6]
```diff
@@ -22,34 +22,28 @@ module.exports = (api) => {
...
-    if (err) {
-      return this.execute(err);
-    }
+    if (err) return this.execute(err);
```
Comment (tshemsedinov) [inline lib/chain.js]: Just shorter syntax

File: lib/chain.js [hunk 7]
```diff
@@ -22,34 +22,28 @@ module.exports = (api) => {
...
+
     op(this.array, item.fn, (err, data) => {
-      if (err) {
-        return this.execute(err);
-      }
+      if (err) return this.execute(err);
```
Comment (tshemsedinov) [inline lib/chain.js]: Just shorter syntax

File: lib/chain.js [hunk 8]
```diff
@@ -22,34 +22,28 @@ module.exports = (api) => {
...
       this.array = data;
       this.execute();
     });
   };
 
   ArrayChain.prototype.fetch = function(fn) {
     this.chain.push({ op: 'fetch', fn });
+    this.execute();
```
Comment (tshemsedinov) [inline lib/chain.js]: Implementing autoexecution of fetch #81

**Comment by tshemsedinov** [comment]
Done and squashed, I am going to merge after CI @nechaido @aqrln

**Comment by tshemsedinov** [comment]
Landed in 798470b

**Pull request closed without merging**